Userland/Libraries/LibPDF/ColorSpace.cpp

```diff
@@ -303,39 +306,66 @@ PDFErrorOr<NonnullRefPtr<ColorSpace>> ICCBasedColorSpace::create(Document* docum
     if (!param.has<NonnullRefPtr<Object>>() || !param.get<NonnullRefPtr<Object>>()->is<StreamObject>())
         return Error { Error::Type::MalformedPDF, "ICCBased color space expects a stream parameter" };
 
-    auto dict = param.get<NonnullRefPtr<Object>>()->cast<StreamObject>()->dict();
-
-    DeprecatedFlyString name;
-    if (!dict->contains(CommonNames::Alternate)) {
-        auto number_of_components = dict->get_value(CommonNames::N).to_int();
-        if (number_of_components == 1)
-            name = CommonNames::DeviceGray;
-        else if (number_of_components == 3)
-            name = CommonNames::DeviceRGB;
-        else if (number_of_components == 4)
-            name = CommonNames::DeviceCMYK;
-        else
-            VERIFY_NOT_REACHED();
-        return ColorSpace::create(name);
-    }
+    auto stream = param.get<NonnullRefPtr<Object>>()->cast<StreamObject>();
+    auto dict = stream->dict();
+
+    auto maybe_profile = Gfx::ICC::Profile::try_load_from_externally_owned_memory(stream->bytes());
+    if (!maybe_profile.is_error())
+        return adopt_ref(*new ICCBasedColorSpace(maybe_profile.release_value()));
+
+    if (dict->contains(CommonNames::Alternate)) {
+        auto alternate_color_space_object = MUST(dict->get_object(document, CommonNames::Alternate));
+        if (alternate_color_space_object->is<NameObject>())
+            return ColorSpace::create(alternate_color_space_object->cast<NameObject>()->name());
 
-    auto alternate_color_space_object = MUST(dict->get_object(document, CommonNames::Alternate));
-    if (alternate_color_space_object->is<NameObject>()) {
-        return ColorSpace::create(alternate_color_space_object->cast<NameObject>()->name());
+        return Error { Error::Type::Internal, "Alternate color spaces in array format are not supported" };
     }
 
-    dbgln("Alternate color spaces in array format not supported yet ");
-    TODO();
+    return Error { Error::Type::MalformedPDF, "Failed to load ICC color space with malformed profile and no alternate" };
 }
 
-Color ICCBasedColorSpace::color(Vector<Value> const&) const
+ICCBasedColorSpace::ICCBasedColorSpace(NonnullRefPtr<Gfx::ICC::Profile> profile)
+    : m_profile(profile)
 {
-    VERIFY_NOT_REACHED();
+}
+
+Color ICCBasedColorSpace::color(Vector<Value> const& arguments) const
+{
+    if (!s_srgb_profile)
```

Comment:
This is alright for now, but eventually the target profile should be passed in to Render, and the GUI should pass in the color space of the window's backing bitmap, and WindowServer should give each bitmap the color space belonging to the monitor the window is on and every color and bitmap should be associated with some color space and painter should do the conversion transparently etc etc etc. Obviously not for this PR 😅

Reply:
Yeahhh I think I'll leave that one for you :smiley: